api/db.py

```diff
@@ -93,6 +93,9 @@ async def update(self, obj):
         if obj.id is None:
             raise ValueError("Cannot update object with no id")
         col = self._get_collection(obj.__class__)
+        node = await col.find_one({'_id': ObjectId(obj.id)})
+        if node:
+            Node.set_created(obj, node.get('created'))
```

Comment:
OK I see, the `created` field is set in `api.main` when making an object if the client didn't send it. Likewise, the `updated` one is also automatically set as a side-effect. But this should be managed by the model rather than `api.main` or `api.db`. So I think we have 2 different issues for `created` and `updated`, we probably should be addressing them separately. Maybe this PR could focus on the `created` field to get it fixed, and then a follow-up would address the `updated` one.

We may need to have some properties set for each field to describe how they should be handled. For example, `created` should not be saved again in the database. So either it has to match exactly what is already there, or an incoming value has to be discarded. I'm not sure which is the best way to do this, maybe the latter with immutable values enforced in Mongo DB would be ideal as the document saved would then have to be what the client sent with the real `created` date.

With the current pipeline, when a node gets updated, normally the data is coming from an existing node with all the fields including `created`. Then the `status` and maybe other fields get updated, but the `created` time should remain the same.  Is that not the case?

Reply:
Yes, it is the case.
As you have suggested using immutable type, I have already tried it with `created` field. I used the below statement to make it immutable.
```
created: Optional[datetime] = Field(default_factory=datetime.utcnow, allow_mutation=False)
```
But it tuned out that It doesn't affect when used with `default_factory`, as `default_factory` is updating the value whenever the object is being saved to the DB.

If I don't put the logic to set `created` field from DB object, `default_factory` will update it automatically. Hence we needed the logic to retrieve the field from the DB and set the field to that value.

Reply:
I believe `allow_mutation` just means that the attribute in the object can't be changed. That's not the same as making it immutable in Mongo DB.

I think we should require the client to either not send `created` or send the real `created` time which should be the same as what is in the database. Then if we make the field in the database immutable, it will cause an error if the value from the client is different. If we can't make the field immutable in Mongo DB (although I believe we can with indexes), then we could add a method to the model to check whether an incoming object contains valid data and compare the immutable fields there instead. If they're not the same then it would raise an exception.